### PR TITLE
KAFKA-12925: adding presfixScan operation for missed implementations

### PR DESCRIPTION
**Pull request by vamossagar12**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java [hunk 1]
```diff
@@ -106,6 +108,8 @@ public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
 
     @Override
     public <PS extends Serializer<P>, P> KeyValueIterator<Bytes, byte[]> prefixScan(final P prefix, final PS prefixKeySerializer) {
+        Objects.requireNonNull(prefix);
+        Objects.requireNonNull(prefixKeySerializer);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: Should we not check for `null` in the outer `MeteredXxxStore`? Should be redundant to check here for a second time.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java [hunk 2]
```diff
@@ -66,6 +68,22 @@ public MemoryNavigableLRUCache(final String name, final int maxCacheSize) {
                 .subMap(from, true, to, true).descendingKeySet().iterator(), treeMap));
     }
 
+    @Override
+    public <PS extends Serializer<P>, P> KeyValueIterator<Bytes, byte[]> prefixScan(final P prefix, final PS prefixKeySerializer) {
+
+        Objects.requireNonNull(prefix);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java]: as above (similar elsewhere)

**Review by mjsax — COMMENTED** [review]
I am not sure if this are all required overwrites.

Can you maybe add an larger test (using TopologyTestDriver) that uses the Processor API, adds a state store (with different variants: caching and/or logging enabled/disabled and RocksDB vs in-memory; also use the old and new Processor API, so we get many different variant of the same test) that tried to get the state store in the `init()`? And maybe also pipe at least one record through the program to access the state store using prefixScan.

**Comment by vamossagar12** [comment]
@mjsax , have added the tests for the combinations that you have mentioned.

**Comment by vamossagar12** [comment]
@mjsax , @ableegoldman  did you get a chance to look at this PR?

**Comment by ableegoldman** [comment]
@vamossagar12 it looks like there are now merge conflicts, can you fix those? I'll try to have a look soon, hopefully @mjsax can as well -- also, can you confirm whether this is/should be a blocker for 3.0 or not?

**Comment by vamossagar12** [comment]
Thanks @ableegoldman  i have fixed the merge conflict. Plz review whenever you get the chance...

**Comment by ableegoldman** [comment]
@vamossagar12 looks like there's a checkstyle failure that's preventing the tests from running, can you fix that as well? Giving this a pass now

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java [hunk 3]
```diff
@@ -26,13 +26,15 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.Iterator;
-import java.util.List;
 import java.util.NavigableMap;
-import java.util.Set;
 import java.util.TreeMap;
+import java.util.List;
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java]: Can you revert the changes in this file?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java [hunk 4]
```diff
@@ -66,6 +68,20 @@ public MemoryNavigableLRUCache(final String name, final int maxCacheSize) {
                 .subMap(from, true, to, true).descendingKeySet().iterator(), treeMap));
     }
 
+    @Override
+    public <PS extends Serializer<P>, P> KeyValueIterator<Bytes, byte[]> prefixScan(final P prefix, final PS prefixKeySerializer) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java:71]: Can you add a test or two for this in `InMemoryLRUCacheStoreTest`? Alternatively, I think you should be able to just collect all the tests in `InMemoryKeyValueStoreTest` and `CachingInMemoryKeyValueStoreTest` and move them over to
`AbstractKeyValueStoreTest` instead. That way you get test coverage for both of those plus a handful of other store classes at once, without having to copy the same test over and over across a bunch of different files.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 5]
```diff
@@ -383,6 +387,1002 @@ public void testDrivingConnectedStateStoreInDifferentProcessorsTopology() {
...
+        assertEquals("value4", results.get(0).value);
+        assertEquals("key2", results.get(1).key);
+        assertEquals("value2", results.get(1).value);
+        assertEquals("key3", results.get(2).key);
+        assertEquals("value3", results.get(2).value);
+
+    }
+
+    @Test
+    public void testPrefixScanPersistentTimestampedStoreNoCachingNoLogging() {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:637]: I'm not sure this is really the right place to test the `prefixScan` functionality for all of these different store types, this test class is really more for making sure the topology itself is all wired up correctly. If you're just trying to test a method on a specific store type, that generally makes sense to do in the test class for that store itself. In other words you don't need to have a separate test here for each underlying store type (eg `PersistentTimestampedStore` or `LruMap`, etc), there are dedicated test classes for that (like `RocksDBTimestampedStoreTest` or `InMemoryLRUCacheStoreTest`)

That said, it sounds like the original bug report uncovered the missing implementations "when accessing the state stores through the processor context" -- which does sound like it could/would be reproduced through a test here. Maybe you can just pick a store type and write a single test that reproduces the issue when run without this patch, and I would consider that sufficient for this.

**Review by ableegoldman — COMMENTED** [review]
I think this looks pretty good (as far as I can tell), but the important thing is going to be the tests, since it's hard to verify just by looking whether we hit every place this is needed. One thing I was wondering was whether we have any tests for this with IQ (Interactive Queries) -- if not, an integration test that tries to run `prefixScan` using IQ should definitely be added

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java (hunk 5 above)
Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:637]: Actually, @mjsax  had asked me to test all possible flavours of the state stores here in this comment:
https://github.com/apache/kafka/pull/10877#pullrequestreview-683669139
I thought this would be the right place to do that. Let me know if it isn't..

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java (hunk 4 above)
Comment (vamossagar12) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java:71]: 1) I guess the reason I didn't add test in `InMemoryLRUCacheStoreTest` was that the key type defined here is Integer and prefixScan like range/reverseRange relies on byte lexicographical ordering and not on the ordering of the key type. So, IMO this test won't yield much for this case. If needed, I can add it.But, I think the tests for reverse and reverseRange are also missing from this class for the same reason or it could just be a miss :) 
2) Regarding the merging of test classes into one, I see that both those classes override the `createKeyValueStore` method differently, so merging them as per me would need some analysis. Is it ok if we create a separate ticket and tackle it there?

**Comment by vamossagar12** [comment]
> @vamossagar12 looks like there's a checkstyle failure that's preventing the tests from running, can you fix that as well? Giving this a pass now

Oh.. sorry about that I believe the checkStyle error should go away now.

**Comment by vamossagar12** [comment]
> I think this looks pretty good (as far as I can tell), but the important thing is going to be the tests, since it's hard to verify just by looking whether we hit every place this is needed. One thing I was wondering was whether we have any tests for this with IQ (Interactive Queries) -- if not, an integration test that tries to run `prefixScan` using IQ should definitely be added

I have added a test for IQ. Some reason, I wasn't able to run the test on my local as I was getting a build failure due to scala test classes. I would watch out for the status of the tests here.

**Comment by ableegoldman** [comment]
> I have added a test for IQ. Some reason, I wasn't able to run the test on my local as I was getting a build failure due to scala test classes. I would watch out for the status of the tests here.

Apparently there was an actual issue with a scala class on trunk, if you pull/rebase you should be able to build locally again. Looks like the test you added did pass though, so nice 👍

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java [hunk 6]
```diff
@@ -845,6 +845,64 @@ public void shouldBeAbleToQueryMapValuesState() throws Exception {
...
+        startKafkaStreamsAndWaitForRunningState(kafkaStreams);
+
+        waitUntilAtLeastNumRecordProcessed(outputTopic, 5);
+
+        final ReadOnlyKeyValueStore<String, Long> myMapStore =
+            IntegrationTestUtils.getStore("queryMapValues", kafkaStreams, keyValueStore());
+
+        int index = 0;
+        final KeyValueIterator<String, Long> range = myMapStore.prefixScan("go", Serdes.String().serializer());
+        while (range.hasNext()) {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:903]: I know this is just how the other tests are doing it, but it's not really an airtight way to validate the expected results...if nothing is returned then we never enter the `while` loop and the test passes, even if we did in fact expect there to be actual output.

The important thing here was just to make sure it didn't throw an exception so it still does that, but it would be good to fix this up maybe in a followup PR

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:637]: Ah, I see, I thought it was trying to test the underlying `prefixScan` functionality itself, but what Matthias said makes sense. They do still feel very out of place here, but I'm not sure I can think of a better place to put them. Maybe just create a new file and put them in a separate test class of their own?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryNavigableLRUCache.java:71]: 1) Could you just treat them as Bytes all the same, and just convert to/from an Integer before putting/getting them from the store? That way you're still just handling Bytes like you are in this test, it just goes through an extra layer of de/serialization. Should be able to more or less copy over the existing tests with just a bit of extra code. Can you try this, in a followup PR?
2) Yes, I was just suggesting to merge them as a possible way to make things easier and do less work, if it's going to be more then please do file a separate ticket for it.

**Review by ableegoldman — APPROVED** [review]
Thanks @vamossagar12 , the actual changes here LGTM. There are still a few things I'd like to address regarding the tests, but the test coverage itself seems good so I think we're ready to just go ahead and merge this in to make code freeze. Let's tackle those remaining items in a followup PR. 

Number of failures, all known flaky and unrelated.

**Comment by ableegoldman** [comment]
Merged to trunk and cherrypicked to 3.0 and 2.8
